Fix some memory bugs in onnx passes

bertmaher opened the pull request:
Running onnx tests with ASAN uncovers several memory errors.  These two are caused by: (1) iterating the uses list of a node after mutation, and (2) accessing the `blocks` attribute of a possibly deleted node.

To reproduce (this is on a CentOS 7 box):
```
DEBUG=1 CFLAGS="-fsanitize=address" CXXFLAGS="-fsanitize=address" USE_LLVM=$(realpath ../llvm-project/install) CMAKE_PREFIX_PATH=$CONDA_PREFIX python setup.py install
LD_PRELOAD=$(realpath /lib64/libasan.so.5) numactl -C3 pytest -v --cov --cov-report xml:test/coverage.xml --cov-append onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11 -s
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63754

Differential Revision: [D30493939](https://our.internmc.facebook.com/intern/diff/D30493939)

cc @gmagogsfm

bertmaher committed:
Fix some memory bugs in onnx passes

[ghstack-poisoned]

malfet commented on torch/csrc/jit/passes/onnx/list_model_parameters.cpp:
```diff
@@ -143,6 +143,8 @@ std::vector<IValue> getParamAttributes(
           auto attrVal = tryInsertConstant(*graph, attr);
           n->output()->replaceAllUsesWith(*attrVal);
           n->destroy();
+          // Don't scan blocks below.
+          continue;
```
Not at all sure, what this code does, but shouldn't we save this to say `nodes_to_destroy` list and iterate over it afterwards? (I.e. why there shouldn't be more nodes like that?)

bertmaher replied:
The continue just picks up with the next node.  If we're destroying this node we don't really care about scanning its sub-blocks, right?

malfet commented on torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:
```diff
@@ -335,6 +336,7 @@ static void PrepareForRemoveMutations(MutationRemover& mr, Block* b) {
         TORCH_INTERNAL_ASSERT(nullptr != newNode);
         node->replaceInput(index, newNode->output());
         input->replaceAllUsesAfterNodeWith(node, newNode->output());
+        goto restart;
```
Yuck. Can't we do this with say do-while loop:
```
  bool needs_restart = false;
  do {
      needs_restart = false;
      ....
      if (cond_met) needs_restart = true;
  } while (needs_restart);

malfet replied:
And I love that `clang-tidy` actually complained about that one :)

bertmaher replied:
I knew I'd get complaints about this but I had to try :-p.  IMO it's just as readable as the boolean but whatever.

jiafatom commented on torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:
```diff
@@ -335,6 +336,7 @@ static void PrepareForRemoveMutations(MutationRemover& mr, Block* b) {
         TORCH_INTERNAL_ASSERT(nullptr != newNode);
         node->replaceInput(index, newNode->output());
         input->replaceAllUsesAfterNodeWith(node, newNode->output());
+        goto restart;
```
This is also a good point. Please don't use `goto`, can we use `break` here?

jiafatom replied:
I see you change the code to remove `goto`, and this `needsRestart` makes sense to me.

bertmaher committed:
Update on "Fix some memory bugs in onnx passes"

Running onnx tests with ASAN uncovers several memory errors.  These two are caused by: (1) iterating the uses list of a node after mutation, and (2) accessing the `blocks` attribute of a possibly deleted node.

To reproduce (this is on a CentOS 7 box):
```
DEBUG=1 CFLAGS="-fsanitize=address" CXXFLAGS="-fsanitize=address" USE_LLVM=$(realpath ../llvm-project/install) CMAKE_PREFIX_PATH=$CONDA_PREFIX python setup.py install
LD_PRELOAD=$(realpath /lib64/libasan.so.5) numactl -C3 pytest -v --cov --cov-report xml:test/coverage.xml --cov-append onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11 -s
```




[ghstack-poisoned]

bertmaher commented:
I caught one more bug (`m->sourceRange().source()` leaves `source` dangling after `sourceRange` is destroyed), and I removed the aesthetically displeasing goto.

jiafatom commented:
LGTM!
nit: please fix clang-format, thanks.

bertmaher committed:
Update on "Fix some memory bugs in onnx passes"

Running onnx tests with ASAN uncovers several memory errors.  These two are caused by: (1) iterating the uses list of a node after mutation, and (2) accessing the `blocks` attribute of a possibly deleted node.

To reproduce (this is on a CentOS 7 box):
```
DEBUG=1 CFLAGS="-fsanitize=address" CXXFLAGS="-fsanitize=address" USE_LLVM=$(realpath ../llvm-project/install) CMAKE_PREFIX_PATH=$CONDA_PREFIX python setup.py install
LD_PRELOAD=$(realpath /lib64/libasan.so.5) numactl -C3 pytest -v --cov --cov-report xml:test/coverage.xml --cov-append onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11 -s
```




[ghstack-poisoned]

bertmaher commented:
clang-format done, thanks for the quick review!

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493939).
